src/core/struct_tree.js

```diff
@@ -67,6 +69,431 @@ class StructTreeRoot {
+    newRefs.push({ ref: catalogRef, data: buffer.join("") });
+
+    const structTreeRoot = new Dict(xref);
+    structTreeRoot.set("Type", Name.get("StructTreeRoot"));
+    const parentTreeRef = xref.getNewTemporaryRef();
+    structTreeRoot.set("ParentTree", parentTreeRef);
+    const kids = [];
+    structTreeRoot.set("K", kids);
+
+    const parentTree = new Dict(xref);
+    const nums = [];
+    parentTree.set("Nums", nums);
+
+    const nextKey = await this.#writeKids({
+      newAnnotationsByPage,
+      structTreeRootRef,
+      kids,
+      nums,
+      xref,
+      pdfManager,
+      newRefs,
+      buffer,
+    });
+    structTreeRoot.set("ParentTreeNextKey", nextKey);
+
+    buffer.length = 0;
+    await writeObject(parentTreeRef, parentTree, buffer, xref);
+    newRefs.push({ ref: parentTreeRef, data: buffer.join("") });
+
+    buffer.length = 0;
+    await writeObject(structTreeRootRef, structTreeRoot, buffer, xref);
+    newRefs.push({ ref: structTreeRootRef, data: buffer.join("") });
+  }
+
+  async canUpdateStructTree({ pdfManager, newAnnotationsByPage }) {
+    if (!this.ref) {
+      warn("Cannot update the struct tree: no root reference.");
+      return false;
+    }
+
+    let nextKey = this.dict.get("ParentTreeNextKey");
+    if (!Number.isInteger(nextKey) || nextKey < 0) {
+      warn("Cannot update the struct tree: invalid next key.");
+      return false;
+    }
+
+    const parentTree = this.dict.get("ParentTree");
+    if (!(parentTree instanceof Dict)) {
+      warn("Cannot update the struct tree: ParentTree isn't a dict.");
+      return false;
+    }
+    const nums = parentTree.get("Nums");
+    if (!Array.isArray(nums)) {
+      warn("Cannot update the struct tree: nums isn't an array.");
+      return false;
+    }
+
+    const { numPages } = pdfManager.catalog;
+    for (const pageIndex of newAnnotationsByPage.keys()) {
+      const { pageDict, ref: pageRef } = await pdfManager.getPage(pageIndex);
+      if (!(pageRef instanceof Ref)) {
+        warn(`Cannot save the struct tree: page ${pageIndex} has no ref.`);
+        return false;
+      }
+      const id = pageDict.get("StructParents");
+      if (!Number.isInteger(id) || id < 0 || id >= numPages) {
+        warn(`Cannot save the struct tree: page ${pageIndex} has no id.`);
+        return false;
+      }
+    }
+
+    const cleanup = map => {
+      for (const elements of map.values()) {
+        for (const element of elements) {
+          delete element.parentTreeId;
+          delete element.structTreeParent;
+        }
+      }
+    };
+
+    let hasNothingToUpdate = true;
+    for (const [pageIndex, elements] of newAnnotationsByPage) {
+      const { pageDict, ref: pageRef } = await pdfManager.getPage(pageIndex);
+      if (!(pageRef instanceof Ref)) {
+        cleanup(newAnnotationsByPage);
+        warn(`Cannot save the struct tree: page ${pageIndex} has no ref.`);
+        return false;
+      }
```

Comment:
Unfortunately I still don't understand under which circumstances the `pageRef` can be invalid here, given the early return in the loop above: https://github.com/mozilla/pdf.js/blob/ebb574e8f4059e035f108c2cfd368616e7c05451/src/core/struct_tree.js#L191-L194

Reply:
I thought I removed it...